### PR TITLE
[#2455] feat(core): Support pluggable catalog operations

### PR DESCRIPTION
**Pull request by FANNG1**
### What changes were proposed in this pull request?
1. add a hacky catalog config 'ops-impl' to specify the custom catalog operation
2. some changes to CatalogOperation
   a. move `entity` from XXCatalogOperation constructor to initialize(), to construct custom catalog operation using reflect  easily.  
   b. move CatalogOperation#initialize() from XXCatalogOperation#newOps() to BaseCatalog#Ops(),  because initialize is the API, it should be called explicitly by the framework not by specific XXCatalogOperation.

### Why are the changes needed?

Fix: #2455 

### Does this PR introduce _any_ user-facing change?
no

### How was this patch tested?
UT and IT

### CODE REVIEW

**Comment by FANNG1** [comment]
@jerryshao @mchades @yuqi1129 , it's ready to review now, please help to review when you are free.

**Comment by FANNG1** [comment]
@TEOTEO520 please help to review when you are free.

File: common/src/main/java/com/datastrato/gravitino/utils/PathUtils.java [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+package com.datastrato.gravitino.utils;
+
+import com.google.common.base.Preconditions;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import org.apache.commons.lang3.StringUtils;
+
+/** Provides Helper methods to process paths. */
+public class PathUtils {
+
```
Comment (yuqi1129) [inline common/src/main/java/com/datastrato/gravitino/utils/PathUtils.java:15]: private constructor

File: catalogs/catalog-hive/src/test/java/com/datastrato/gravitino/catalog/hive/integration/test/CatalogHiveIT.java [hunk 2]
```diff
@@ -1357,4 +1360,45 @@ public void testPurgeHiveExternalTable() throws TException, InterruptedException
     Assertions.assertTrue(
         hdfs.listStatus(tableDirectory).length > 0, "The table should not be empty");
   }
+
+  @Test
+  @EnabledIf("isDeployMode")
```
Comment (yuqi1129) [inline catalogs/catalog-hive/src/test/java/com/datastrato/gravitino/catalog/hive/integration/test/CatalogHiveIT.java:1365]: Why do we only enable it in deploy mode?

File: core/src/main/java/com/datastrato/gravitino/catalog/BaseCatalog.java [hunk 3]
```diff
@@ -101,6 +106,24 @@ public CatalogOperations ops() {
...
 
+  private CatalogOperations createOps(Map<String, String> conf) {
+    String customCatalogOperationClass = conf.get(CATALOG_OPERATION_CLASS_NAME);
+    return Optional.ofNullable(customCatalogOperationClass)
+        .map(className -> loadCustomerOps(className))
+        .orElse(newOps(conf));
+  }
+
+  private CatalogOperations loadCustomerOps(String className) {
+    ClassLoader classLoader = Thread.currentThread().getContextClassLoader();
```
[the remark this answers is not in the document]
Comment (FANNG1) [reply core/src/main/java/com/datastrato/gravitino/catalog/BaseCatalog.java:117]: yes, couldn't find package if not sepcifying current classloader

File: catalogs/catalog-hadoop/src/main/java/com/datastrato/gravitino/catalog/hadoop/HadoopCatalogOperations.java [hunk 4]
```diff
@@ -77,17 +77,17 @@ public class HadoopCatalogOperations implements CatalogOperations, SupportsSchem
...
 
-  public HadoopCatalogOperations(CatalogEntity entity) {
-    this(entity, GravitinoEnv.getInstance().entityStore());
+  public HadoopCatalogOperations() {
+    this(GravitinoEnv.getInstance().entityStore());
   }
 
   @Override
-  public void initialize(Map<String, String> config) throws RuntimeException {
+  public void initialize(Map<String, String> config, CatalogEntity entity) throws RuntimeException {
```
Comment (SteNicholas) [inline catalogs/catalog-hadoop/src/main/java/com/datastrato/gravitino/catalog/hadoop/HadoopCatalogOperations.java:89]: Could the `CatalogOperations` interface add `setCatalogEntity` method?

File: core/src/main/java/com/datastrato/gravitino/catalog/CatalogManager.java [hunk 5]
```diff
@@ -581,6 +593,15 @@ private String buildConfPath(String provider) {
     return String.join(File.separator, gravitinoHome, "catalogs", provider, "conf");
   }
 
+  private Optional<String> buildCustomCatalogOperationPath(Map<String, String> catalogProperties) {
+    String className = catalogProperties.get(Catalog.CATALOG_OPERATION_CLASS_NAME);
+    String path = catalogProperties.get(Catalog.CATALOG_OPERATION_CLASS_PATH);
+    if (StringUtils.isNotBlank(className) && StringUtils.isNotBlank(path)) {
```
Comment (SteNicholas) [inline core/src/main/java/com/datastrato/gravitino/catalog/CatalogManager.java:599]: Could the `path` have the default value?

Comment (FANNG1) [reply core/src/main/java/com/datastrato/gravitino/catalog/CatalogManager.java:599]: the default action is not adding extra path to catalog classpaths.

File: catalogs/catalog-hadoop/src/main/java/com/datastrato/gravitino/catalog/hadoop/HadoopCatalogOperations.java (hunk 4 above)
Comment (FANNG1) [reply catalogs/catalog-hadoop/src/main/java/com/datastrato/gravitino/catalog/hadoop/HadoopCatalogOperations.java:89]: this may make CatalogOperations API more complicated, @jerryshao what do you think?

File: common/src/main/java/com/datastrato/gravitino/utils/PathUtils.java [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+  /**
+   * Helper methods to transform path to absolutePath, if path is relative, it's supposed to under
+   * gravitinoHome directory.
+   *
+   * @param path The path to transform.
+   * @param gravitinoHome Gravitino home directory.
+   * @return The absolute path.
+   */
+  public static String getAbsolutePath(String path, String gravitinoHome) {
```
[the remark this answers is not in the document]
Comment (FANNG1) [reply common/src/main/java/com/datastrato/gravitino/utils/PathUtils.java:24]: "GRAVITINO_HOME" is not set in UnitTest

File: api/src/main/java/com/datastrato/gravitino/Catalog.java [hunk 7]
```diff
@@ -40,6 +40,22 @@ enum Type {
...
   String PROPERTY_PACKAGE = "package";
 
+  /**
+   * Custom catalog operation class name. It is recommended to extend current Gravitino catalog
+   * operations like HiveCatalogOperation, and override interfaces like purgeTable, and
+   * listPartition which should be hacked, but this is not a must. It only needs to implement
+   * CatalogOperations to initialize itself, implement SupportsSchemas to do schema ops, and
+   * TableCatalog to do table ops if necessary.
+   */
+  String CATALOG_OPERATION_CLASS_NAME = "catalog.operation.custom.class-name";
```
[the remark this answers is not in the document]
Comment (FANNG1) [reply api/src/main/java/com/datastrato/gravitino/Catalog.java:50]: thanks for suggesting, I don't like the name too, I'd like to hear more reviews's idea.

File: common/src/main/java/com/datastrato/gravitino/utils/PathUtils.java (hunk 1 above)
Comment (FANNG1) [reply common/src/main/java/com/datastrato/gravitino/utils/PathUtils.java:15]: done

File: catalogs/catalog-hive/src/test/java/com/datastrato/gravitino/catalog/hive/integration/test/CatalogHiveIT.java (hunk 2 above)
Comment (FANNG1) [reply catalogs/catalog-hive/src/test/java/com/datastrato/gravitino/catalog/hive/integration/test/CatalogHiveIT.java:1365]: Couldn't run under embedded mode, because Iceberg package may not build yet.

File: core/src/test/java/com/datastrato/gravitino/catalog/TestBaseCatalog.java [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+import com.datastrato.gravitino.meta.CatalogEntity;
+import com.google.common.collect.ImmutableMap;
+import org.junit.jupiter.api.Assertions;
+import org.junit.jupiter.api.Test;
+import org.mockito.Mockito;
+
+public class TestBaseCatalog {
+
+  @Test
+  void testCustomCatalogOperations() {
```
[the remark this answers is not in the document]
Comment (FANNG1) [reply core/src/test/java/com/datastrato/gravitino/catalog/TestBaseCatalog.java:20]: classpath test is coverd in integrate test in CatalogHiveIT#testCustomCatalogOperations.  It's hard to add a UT for classpath, maybe we could add test for `buildCustomCatalogOperationPath` in `CatalogManager`, but seems not much value. what do you think?

File: catalogs/catalog-hive/src/test/java/com/datastrato/gravitino/catalog/hive/integration/test/CatalogHiveIT.java (hunk 2 above)
Comment (yuqi1129) [reply catalogs/catalog-hive/src/test/java/com/datastrato/gravitino/catalog/hive/integration/test/CatalogHiveIT.java:1366]: Can we make the ut depend on task.jar?

Comment (jerryshao) [reply catalogs/catalog-hive/src/test/java/com/datastrato/gravitino/catalog/hive/integration/test/CatalogHiveIT.java:1366]: This comment is not addressed. And the way you did is tricky, you need to fix it not bypass it. @FANNG1

File: catalogs/catalog-hadoop/src/main/java/com/datastrato/gravitino/catalog/hadoop/HadoopCatalogOperations.java (hunk 4 above)
Comment (jerryshao) [reply catalogs/catalog-hadoop/src/main/java/com/datastrato/gravitino/catalog/hadoop/HadoopCatalogOperations.java:89]: What's the purpose of this change?

File: catalogs/catalog-hive/src/test/java/com/datastrato/gravitino/catalog/hive/integration/test/CatalogHiveIT.java (hunk 2 above)
Comment (FANNG1) [reply catalogs/catalog-hive/src/test/java/com/datastrato/gravitino/catalog/hive/integration/test/CatalogHiveIT.java:1366]: Seems there is no strong necessary to make it run in embedded mode. The main purpose of this test is to load a custom catalog operations from external class path, run in deploy mode is enough.

File: catalogs/catalog-hadoop/src/main/java/com/datastrato/gravitino/catalog/hadoop/HadoopCatalogOperations.java (hunk 4 above)
Comment (FANNG1) [reply catalogs/catalog-hadoop/src/main/java/com/datastrato/gravitino/catalog/hadoop/HadoopCatalogOperations.java:89]: `CatalogEntity` is passed to every specific XXCatalogOperation in class constructor, like `HiveCatalogOperations(CatalogEntity entity)`, but there is no API interface to define it, how to pass `CatalogEntity` to a custom catalog operation? I propose to pass though initialize interface explicitly.

File: api/src/main/java/com/datastrato/gravitino/Catalog.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (jerryshao) [reply api/src/main/java/com/datastrato/gravitino/Catalog.java:50]: Basically I have several comments about this config:

1. Make this catalog property an undocumented property, don't expose here. This property should not be used by most of the users, it's just a secret hole for particular users, we don't encourage this also. So remove the definition here and put it into the property metadata only.
2. Change the name "catalog.operation.custom.class-name" to "ops-impl" for simplicity.
3. Only this configuration "ops-impl" is enough, don't need this "catalog.operation.custom.classpath". We assume users who want to use this feature should know and put their code/jar within catalog classpath. We don't want to support too much to load it from different path.

Basically, this is a property that I don't want to expose to all the users, it's quite hacky. So we should keep the changes as invisible as we can.

File: catalogs/catalog-hadoop/src/main/java/com/datastrato/gravitino/catalog/hadoop/HadoopCatalogOperations.java (hunk 4 above)
Comment (jerryshao) [reply catalogs/catalog-hadoop/src/main/java/com/datastrato/gravitino/catalog/hadoop/HadoopCatalogOperations.java:89]: Things doesn't change even if you maven this to `initialize`, `CatalogEntity` is still not a public API we defined currently.

Comment (FANNG1) [reply catalogs/catalog-hadoop/src/main/java/com/datastrato/gravitino/catalog/hadoop/HadoopCatalogOperations.java:89]: do you have some suggestions?

File: api/src/main/java/com/datastrato/gravitino/Catalog.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (FANNG1) [reply api/src/main/java/com/datastrato/gravitino/Catalog.java:50]: ok,  @SteNicholas what do you think?

File: catalogs/catalog-hadoop/src/main/java/com/datastrato/gravitino/catalog/hadoop/HadoopCatalogOperations.java (hunk 4 above)
Comment (jerryshao) [reply catalogs/catalog-hadoop/src/main/java/com/datastrato/gravitino/catalog/hadoop/HadoopCatalogOperations.java:89]: I don't have suggestions, I just want to know the purpose of this change, this seems an unrelated change. If there's a benefit I think it is OK to change. If not, then not necessary to change it.

Comment (FANNG1) [reply catalogs/catalog-hadoop/src/main/java/com/datastrato/gravitino/catalog/hadoop/HadoopCatalogOperations.java:89]: The main purpose of moving passing `CatalogEntity` to initialize is to make construct a custom catalog operation using reflect easily.

Comment (jerryshao) [reply catalogs/catalog-hadoop/src/main/java/com/datastrato/gravitino/catalog/hadoop/HadoopCatalogOperations.java:89]: OK, I see.

File: core/src/main/java/com/datastrato/gravitino/catalog/BaseCatalog.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (FANNG1) [reply core/src/main/java/com/datastrato/gravitino/catalog/BaseCatalog.java]: remove extra class path related code

File: catalogs/catalog-hive/src/test/java/com/datastrato/gravitino/catalog/hive/integration/test/CatalogHiveIT.java (hunk 2 above)
Comment (FANNG1) [reply catalogs/catalog-hive/src/test/java/com/datastrato/gravitino/catalog/hive/integration/test/CatalogHiveIT.java]: no need to load custom catalog operation from external class path, remove the limit now.

File: core/src/main/java/com/datastrato/gravitino/catalog/CatalogManager.java (hunk 5 above)
Comment (FANNG1) [reply core/src/main/java/com/datastrato/gravitino/catalog/CatalogManager.java]: code removed

**Comment by FANNG1** [comment]
@jerryshao , comment are addressed, please review again

File: core/src/main/java/com/datastrato/gravitino/catalog/BaseCatalog.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (FANNG1) [reply core/src/main/java/com/datastrato/gravitino/catalog/BaseCatalog.java:121]: We must specify the classloader even if the custom class package is placed under the corresponding catalogs directory.

File: api/src/main/java/com/datastrato/gravitino/Catalog.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (SteNicholas) [reply api/src/main/java/com/datastrato/gravitino/Catalog.java]: +1.

File: catalogs/catalog-hadoop/src/main/java/com/datastrato/gravitino/catalog/hadoop/HadoopCatalogOperations.java [hunk 9]
```diff
@@ -77,17 +77,17 @@ public class HadoopCatalogOperations implements CatalogOperations, SupportsSchem
...
+    this.entity = entity;
```
Comment (SteNicholas) [inline catalogs/catalog-hadoop/src/main/java/com/datastrato/gravitino/catalog/hadoop/HadoopCatalogOperations.java:90]: Could we introduce `BaseCatalogOperations` to extract common behavior like this?

File: core/src/main/java/com/datastrato/gravitino/catalog/BaseCatalog.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (FANNG1) [reply core/src/main/java/com/datastrato/gravitino/catalog/BaseCatalog.java:121]: Class.forName use classloader of the caller class (BaseCatalog.class), it's global classloader not the catalog specific classloader, so we must specify the classloader explicitly.

**Comment by mchades** [comment]
Have you tested whether custom CatalogOperations can successfully pass end-to-end tests?

File: core/src/main/java/com/datastrato/gravitino/catalog/BaseCatalog.java [hunk 10]
```diff
@@ -106,6 +115,27 @@ public CatalogOperations ops() {
...
+    try {
+      // Class.forName use classloader of the caller class (BaseCatalog.class), it's global
+      // classloader not the catalog specific classloader, so we must specify the classloader
+      // explicitly.
+      return (CatalogOperations)
+          Class.forName(className, true, classLoader).getDeclaredConstructor().newInstance();
```
[the remark this answers is not in the document]
Comment (FANNG1) [reply core/src/main/java/com/datastrato/gravitino/catalog/BaseCatalog.java:132]: I think there's no necessary to check `SupportsSchemas` or `CatalogOperations`

**Comment by FANNG1** [comment]
> Have you tested whether custom CatalogOperations can successfully pass end-to-end tests?

good point, I could run pass HiveCatalogIT using custom catalog operations in local machine.

File: core/src/main/java/com/datastrato/gravitino/catalog/BaseCatalog.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (mchades) [reply core/src/main/java/com/datastrato/gravitino/catalog/BaseCatalog.java:132]: Why is it unnecssary? If these checks are not performed, the custom operation will not be usable.

File: core/src/main/java/com/datastrato/gravitino/catalog/BaseCatalog.java [hunk 11]
```diff
@@ -27,6 +30,11 @@
  */
 public abstract class BaseCatalog<T extends BaseCatalog>
     implements Catalog, CatalogProvider, HasPropertyMetadata {
+  private static final Logger LOG = LoggerFactory.getLogger(BaseCatalog.class);
+
+  // A hack way to inject custom operation to Gravitino, the object you used is not stable, don't
+  // use it unless you know what you are doing.
+  @VisibleForTesting public static final String CATALOG_OPERATION_IMPL = "ops-impl";
```
Comment (jerryshao) [inline core/src/main/java/com/datastrato/gravitino/catalog/BaseCatalog.java:37]: If this is visible for test, I think package private is enough, no need to be public.

Comment (FANNG1) [reply core/src/main/java/com/datastrato/gravitino/catalog/BaseCatalog.java:37]: this's is public because it's used by `CatalogHiveIT`

File: core/src/main/java/com/datastrato/gravitino/catalog/BaseCatalog.java [hunk 12]
```diff
@@ -106,6 +115,27 @@ public CatalogOperations ops() {
...
   }
 
+  private CatalogOperations createOps(Map<String, String> conf) {
+    String customCatalogOperationClass = conf.get(CATALOG_OPERATION_IMPL);
+    return Optional.ofNullable(customCatalogOperationClass)
+        .map(className -> loadCustomerOps(className))
+        .orElse(newOps(conf));
+  }
+
+  private CatalogOperations loadCustomerOps(String className) {
```
[the remark this answers is not in the document]
Comment (FANNG1) [reply core/src/main/java/com/datastrato/gravitino/catalog/BaseCatalog.java:125]: ok

[the remark this answers is not in the document]
Comment (FANNG1) [reply core/src/main/java/com/datastrato/gravitino/catalog/BaseCatalog.java]: fixed, please review again

File: core/src/main/java/com/datastrato/gravitino/catalog/BaseCatalog.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (mchades) [reply core/src/main/java/com/datastrato/gravitino/catalog/BaseCatalog.java:132]: Alright, I can convince myself in this way: making it difficult to use is also a strategy, and this strategy implies that we do not recommend users to use it.

File: core/src/main/java/com/datastrato/gravitino/catalog/BaseCatalog.java (hunk 11 above)
Comment (jerryshao) [reply core/src/main/java/com/datastrato/gravitino/catalog/BaseCatalog.java:37]: I think you need to put the property into catalog property metadata to manage the properties altogether.
